### ISSUE TITLE
Preferences dialog lacks consistent UI style.

### ISSUE THREAD

**Issue by pixelzoom** [phetsims/joist#842]
In #837, Options were folded into the Preferences dialog.  In general, the content for the Options dialog was not changed, it was simply passed to the Preferences dialog.  The end result is that there are now different UI styles in the Preferences dialog, a bit of a hodge podge.

For example, see Gas Properties. Projector Mode (handled by Preferences dialog) is set using an A/B switch, with a nice description of the preference, and looks like this:

<img width="750" alt="screenshot_1818" src="https://user-images.githubusercontent.com/3046552/186278331-374de523-7697-4e3b-be4d-dff3791ee7d4.png">
 
Options specific to Gas Properties used a checkbox instead of an A/B switch, and has no description. It looks like this:

<img width="744" alt="screenshot_1819" src="https://user-images.githubusercontent.com/3046552/186278483-f08e37ac-108a-4145-9596-9f16704cc4cd.png">

A style guide and some common-code UI components are needed.

PhET can certainly live with this for now, because it's usable.  If we didn't have to worry about PhET-iO, this issue would be a "polish".  But changing to a consistent UI style in the future will break the PhET-iO API, require migration rules, etc.

**Comment by terracoda** [comment]
Great points, @pixelzoom. I think it might be better for me to comment in the other issue where we are discussing wording of the "inclusive features".

Perhaps a temporary fix is using a "Sim-specifc Options" subheading here for these types of checkboxes that are being migrated directly from the PhET Menu to the Preferences Menu.

**Comment by terracoda** [comment]
This issue is also related to #838

**Comment by amanda-phet** [comment]
Related to this issue, we discovered that the right-aligned toggle switches can look too close to the next column and leads to a very confusing interface. 

![image (1)](https://user-images.githubusercontent.com/7852916/199084280-7b8b7522-8915-4587-9dd3-ca4f1241b874.png)

A left-aligned toggle switch seems like it would work a lot better, and then the toggles will be aligned in every preferences dialog and their placement won't be all over the place.

**Comment by amanda-phet** [comment]
Discussed a unified design for the preferences dialog 11/17/22

* When should we use toggles vs. checkboxes? 
* How should we display spinners?
* What about aqua radio buttons? 

Current toggle paradigm works well in a single-column format with right-aligned toggles. 
* Looking at MacOS and Chrome preferences, they have "tabs" (categories) on the left side, and a scrollable single column of preferences (with right-aligned toggles). 
* Why don't we have a scrollable preferences dialog? Because we wanted everything visible on the screen and not have something that is hard to find. We'd need to build something in scenery to allow scrolling (scrollPane?) if we want this in the preferences dialog. 
* Scrolling might be annoying if it dramatically changes the size of the scroll bar. If there is not enough content to activate the scrollbar we'd want it to not show up. Could possibly have a scrollbar that appears and disappears depending on what is checked. 
* Having one column can be hard because certain content may be hidden lower. 
* scrolling would need to work with focus traversal 

Another inconsistency: 
* Sometimes a toggle shows/hides content below, but sometimes it grays out content below. 

If a tab has multiple columns, should we make the column widths consistent? Rather than base it on the contents of the column? 
* would make to make sure these column widths are responsive, since the preferences dialog can sometimes be really wide (if locales are available). 

Different tabs seem to have different types of control 
* simulation tab is totally custom 
* audio tab is pre-set as two columns
* there needs to be logic for the layout that is responsive and consistent across all sims with preferences dialogs; a template that anyone could put their content into
* there should also be reusable components that give us the look and information that we want (e.g. adding descriptions to preferences) 

Immediate concerns: 
* lots of sims are on deck to be published with a preferences dialog that is a little wonky. For now, this is ok by @kathy-phet  and these sims won't be blocked. 
* once we have a unified design, will we have the resources to return to those sims and re-publish them with an updated dialog? @kathy-phet says likely yes, since many of these sims are PhET-iO sims. 
* Some sims have the "options" content being moved over to the "simulation" tab, and don't match the style in other dialogs because there isn't an established design. 
* @jessegreenberg made a "preferences toggle button" that is a toggle switch under the hood with role=switch. Checkboxes in preferences dialog are special and not the same as the ones in sims- there was a reason for this but we can't remember, possibly related to voicing. 

Next steps: 
* @marlitas could create reusable components that devs could use to put in simulation tabs based on the look and feel of the audio tab preferences. (could these have the format of "name" and "description" ?) 
* @marlitas should dig in to the current preferences components and untangle/document what these special components are (checkbox, toggle button, maybe others). 
* might come back to the idea of scrolling/single column as well 
* explore the idea of each column having a faint box around them to help define the left and right columns 
* **need to decide if we are going in the two-column or one-column direction**
* @catherinecarter will take a stab at doing some mockups! @arouinfar will pair and be a thought partner.

**Comment by terracoda** [comment]
Regarding the first action item, I would like for us to use a consistent language for things that require words, i.e. the text that can be recognized by screen readers or delivered via the Voicing feature.

"description" might be too general as everything from the Sim's name in the H1 to a context response that fires as changes happen is a description. It might be useful to call an interactive object's description "help text" if that is the function of the description.

**Comment by pixelzoom** [comment]
@marlitas After the above changes, I'm seeing something very odd in PreferencesToggleSwitch, and therefore in calculus-grapher ValueControl.  PreferencesToggleSwitch is now unrelated to ToggleSwitch -- are you planning to rename PreferencesToggleSwitch?

**Comment by marlitas** [comment]
@pixelzoom Yes the rename is happening today. Ran out of time last night. It has now been refactored to take any control in the upper right corner (or wherever we decide in terms of design in the future). This has been done to make preferences components more consistent among sims and easier for developers to use. What was the odd behavior you were seeing? I want to make sure I refactored all cases out properly.

**Comment by pixelzoom** [comment]
Two things:

* It's odd that all subcomponents of PreferencesToggleSwitch are optional. Shouldn't `controlNode` be required? 

* You have a lot of SelfOptions fields that have type `null | Node`. It guess that works. But it's preferrable to define then as `Node`, then StrictOmit them in the optionize call.  

```diff
-   labelNode?: null | Node;
+   labelNode?: Node;

- const options = optionize<PreferencesToggleSwitchOptions, SelfOptions, NodeOptions>()( {
+ const options = optionize<PreferencesToggleSwitchOptions, StrictOmit<SelfOptions, 'labelNode'>, NodeOptions>()( { 
```

**Comment by marlitas** [comment]
> It's odd that all subcomponents of PreferencesToggleSwitch are optional. Shouldn't controlNode at least be required?

The hope is that his component will work for all the different use cases a preferences dialog may need. For example, a component may need a title and a control, but there are also moments when it may only need a title and description, or just a title, or a description and a control. Making all of these optional will provide flexibility while maintaining a consistent UI across the different usages. The name `PreferencesControl` is 100% up for discussion as it does not fully encompass these usages, but it felt like a step in the right direction.

**Comment by marlitas** [comment]
I updated NumberPlayPreferencesNode to use the new PreferencesControl API. @chrisklus the commits are above if you want to take a look and reach out to me for any questions or concerns.

**Comment by chrisklus** [comment]
@marlitas and I reviewed the changes in Number Play together. Thanks! Back to @marlitas.

**Comment by marlitas** [comment]
Reopening, as the work for preferences isn't done here, but it is done for Number Play. :-)

**Comment by chrisklus** [comment]
Ope, my bad! Forgot this wasn't in a Number Play repo.

**Comment by pixelzoom** [comment]
At 4/21/25 developer meeting, we discussed this issue.  Summary:

* Because Preferences is now included in all sims, this issue has become more important and deserves to be elevated in priority. PhET is building a large amount of technical debt.
* Because of the prominence of Preferences in PhET Studio, this issue has become more important and deserves to be elevated in priority.  The current "look" of Preferences is not up to PhET UI design standards.
* The current layout of many Preferences controls makes them confusing.
* Joist's `PreferencesControl` is inadequate for addressing the needs of Preferences. PhET needs more reusable code for specific types of controls (AB switch, radio buttons, etc.)
* Some sims have addressed the deficiencies by rolling their own Preferences controls. This is building still more technical debt, and should be avoided.
* Addressing this issue will require input from both designers and developers.

**Comment by marlitas** [comment]
Documenting a preferences design question that came up in a Number Pairs dev test: https://github.com/phetsims/joist/issues/1035
